zebra-consensus/src/primitives/halo2.rs

```diff
@@ -0,0 +1,281 @@
+};
+
+use futures::future::{ready, Ready};
+use once_cell::sync::Lazy;
+use orchard::circuit::VerifyingKey;
+use rand::{thread_rng, CryptoRng, RngCore};
+use thiserror::Error;
+use tokio::sync::broadcast::{channel, error::RecvError, Sender};
+use tower::{util::ServiceFn, Service};
+use tower_batch::{Batch, BatchControl};
+use tower_fallback::Fallback;
+
+#[cfg(test)]
+mod tests;
+
+lazy_static::lazy_static! {
+    pub static ref VERIFYING_KEY: VerifyingKey = VerifyingKey::build();
+}
+
+// === TEMPORARY BATCH HALO2 SUBSTITUTE ===
+//
+// These types are meant to be API compatible with the batch verification APIs
+// in bellman::groth16::batch, reddsa::batch, redjubjub::batch, and
+// ed25519-zebra::batch. Once Halo2 batch proof verification math and
+// implementation is available, this code can be replaced with that.
+
+/// A Halo2 verification item, used as the request type of the service.
+#[derive(Clone, Debug)]
+pub struct Item {
+    instances: Vec<orchard::circuit::Instance>,
+    proof: orchard::circuit::Proof,
+}
+
+impl Item {
+    /// Perform non-batched verification of this `Item`.
+    ///
+    /// This is useful (in combination with `Item::clone`) for implementing
+    /// fallback logic when batch verification fails.
+    pub fn verify_single(&self, vk: &VerifyingKey) -> Result<(), halo2::plonk::Error> {
+        self.proof.verify(vk, &self.instances[..])
+    }
+}
+
+#[derive(Default)]
+pub struct BatchVerifier {
+    queue: Vec<Item>,
+}
+
+impl BatchVerifier {
+    pub fn queue(&mut self, item: Item) {
+        self.queue.push(item);
+    }
+
+    pub fn verify<R: RngCore + CryptoRng>(
+        self,
+        _rng: R,
+        vk: &VerifyingKey,
+    ) -> Result<(), halo2::plonk::Error> {
+        for item in self.queue {
+            item.verify_single(vk)?;
+        }
+
+        Ok(())
+    }
+}
+
+// === END TEMPORARY BATCH HALO2 SUBSTITUTE ===
+
+impl From<zebra_chain::orchard::ShieldedData> for Item {
+    fn from(shielded_data: zebra_chain::orchard::ShieldedData) -> Item {
+        use orchard::{circuit, note, primitives::redpallas, tree, value};
+
+        let anchor = tree::Anchor::from_bytes(shielded_data.shared_anchor.into()).unwrap();
+
+        let enable_spend = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_SPENDS);
+        let enable_output = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_OUTPUTS);
+
+        let instances = shielded_data
+            .actions()
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .unwrap(),
```

Comment:
Out of scope for this PR, but it would be cool to have `impl<T> TryFrom<redpallas::VerificationKeyBytes<T>> for redpallas::VerificationKey<T>` or something, just to optimize the conversions here :laughing:

Reply:
It would yeah 😻

Reply:
Well, that exists, the problem is these are two different implementations of `redpallas`: the one in `zebra_chain` and the fork of `redjubjub` that  is re-exported by `orchard`, and looks almost the exact same 😅